File: core/x/javaapi/api/src/main/java/viaduct/java/api/resolvers/ConnectionResolverBase.java
package viaduct.java.api.resolvers;

import viaduct.java.api.context.ConnectionFieldExecutionContext;
import viaduct.java.api.types.Connection;
import viaduct.java.api.types.ConnectionArguments;
import viaduct.java.api.types.GraphQLObject;
import viaduct.java.api.types.Query;

/**
 * Base interface for connection field resolver implementations in the Java Tenant API.
 *
 * <p>Java equivalent of Kotlin's {@code ConnectionResolverBase}. Narrows {@link FieldResolverBase}
 * so a connection resolver's arguments are typed as {@link ConnectionArguments} and its output as a
 * {@link Connection}. Generated connection resolver base classes implement this interface.
 *
 * @param <T> the return type of the resolver (the connection GraphQL type)
 * @param <O> the parent object type containing this field
 * @param <Q> the Query root type
 * @param <A> the connection-arguments type for this field
 * @param <R> the connection output type for the selections
 */
public interface ConnectionResolverBase<
        T,
        O extends GraphQLObject,
        Q extends Query,
        A extends ConnectionArguments,
        R extends Connection<?, ?>>
    extends FieldResolverBase<T, O, Q, A, R> {

  /** Context type alias for connection resolvers, narrowing to {@link ConnectionArguments}. */
  interface Context<
          O extends GraphQLObject,
          Q extends Query,
          A extends ConnectionArguments,
          R extends Connection<?, ?>>
      extends ConnectionFieldExecutionContext<O, Q, A, R> {}
}

File: core/tenant/api/src/main/kotlin/viaduct/api/ConnectionResolverBase.kt
package viaduct.api

import viaduct.api.types.Connection
import viaduct.api.types.ConnectionArguments
import viaduct.api.types.Object
import viaduct.api.types.Query
import viaduct.apiannotations.TypeInferenceApi

/**
 * Specialized resolver base for connection field resolvers. Extends [FieldResolverBase] with the
 * additional constraint that A is a [ConnectionArguments], exposing pagination utilities.
 *
 * R's upper bound is `Connection<*, *>?` so the generator can emit nullable connection return
 * types (e.g. `FooConnection?`) without tripping the bound check.
 *
 * This interface exists to support the testing framework only. Other framework code should depend
 * on [ResolverBase] instead. Generated base classes implement both directly.
 */
@TypeInferenceApi
interface ConnectionResolverBase<O : Object, Q : Query, A : ConnectionArguments, R : Connection<*, *>?> :
    FieldResolverBase<O, Q, A, R>
